lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -451,3 +454,77 @@ void SupplementaryOutputPathsComputer::deriveModulePathParameters(
   mainOutputIfUsable =
       canUseMainOutputForModule && !OutputFiles.empty() ? OutputFiles[0] : "";
 }
+
+static SupplementaryOutputPaths
+createFromTypeToPathMap(const TypeToPathMap *map) {
+  SupplementaryOutputPaths paths;
+  if (!map)
+    return paths;
+  static const std::vector<std::pair<types::ID, std::string &>> typesAndStrings{
+      {types::TY_ObjCHeader, paths.ObjCHeaderOutputPath},
+      {types::TY_SwiftModuleFile, paths.ModuleOutputPath},
+      {types::TY_SwiftModuleDocFile, paths.ModuleDocOutputPath},
+      {types::TY_Dependencies, paths.DependenciesFilePath},
+      {types::TY_SwiftDeps, paths.ReferenceDependenciesFilePath},
+      {types::TY_SerializedDiagnostics, paths.SerializedDiagnosticsPath},
+      {types::TY_ModuleTrace, paths.LoadedModuleTracePath},
+      {types::TY_TBD, paths.TBDPath}};
+  for (auto typeAndString : typesAndStrings) {
+    auto const out = map->find(typeAndString.first);
+    typeAndString.second = out == map->end() ? "" : out->second;
+  }
+  return paths;
+}
+
+Optional<std::vector<SupplementaryOutputPaths>>
+SupplementaryOutputPathsComputer::readSupplementaryOutputFileMap() const {
+  if (Arg *A = Args.getLastArg(options::OPT_emit_objc_header_path,
+                               options::OPT_emit_module_path,
+                               options::OPT_emit_module_doc_path,
+                               options::OPT_emit_dependencies_path,
+                               options::OPT_emit_reference_dependencies_path,
+                               options::OPT_serialize_diagnostics_path,
+                               options::OPT_emit_loaded_module_trace_path,
+                               options::OPT_emit_tbd_path)) {
+    // Since the table is only used for a diagnostic, spend the time to recreate
+    // it and save the complexity of passing it in.
+    std::unique_ptr<llvm::opt::OptTable> opts(createSwiftOptTable());
```

Comment:
Please don't. It's okay to use string constants for command-line arguments in diagnostics, especially if it's a diagnostic that will only show up for people writing tests.

Reply:
OK, but I don't understand why. Is there some performance criticality here I don't know about?

Reply:
In this case it's more code complexity. This is jumping through hoops to get something that's ultimately a static string, and on top of that it's a string that's unlikely to ever change.

Reply:
OK, changed!